github: adding a pull request template

bobafetthotmail opened the pull request:
This text is used by GitHub to remind important things to people posting PRs through the GitHub's web interface. See here for more information https://github.com/blog/2111-issue-and-pull-request-templates

Loosely inspired by OpenWRT package feed's own pull request template
https://github.com/openwrt/packages/blob/master/.github/pull_request_template

Signed-off-by: Alberto Bursi <alberto.bursi@outlook.it>

bobafetthotmail commented:
feel free to add more requirements to the list in this template.

pepe2k commented:
This is really great!

I would also add/mention:
- preferred approach is to use separate branch for every PR
- short tutorial how to rebase and update PR (I see that many people have problems with that when they are asked for changes)

Cheers

miltador commented:
@pepe2k, your points are good. But I think it would be also nice to have CONTRIBUTING.md file in the root of the project near readme, in which there will be described your points and maybe also other dedicated to contributing info (may be from LEDE Wiki). GitHub has imtegration of this file and points user to read it before creating PR.

diizzyy commented:
I managed to help someone on IRC using this quick and ugly hack...
It's far from optimal but at least this person managed to do a successful pull request :-)
http://paste.ubuntu.com/23472999/

bobafetthotmail commented:
@miltador the openwrt packages feed has a CONTRIBUTING.md file we can also mostly copy. https://github.com/openwrt/packages/blob/master/CONTRIBUTING.md

Most of other stuff proposed is good for a wiki page, not a short reminder, thanks dizzyyy for that pastebin. Will make a wiki page soon-ish.

wongsyrone commented:
I think `<device-name>` should be `<arch-name>`

EricLuehrsen commented:
Maybe it could be something more direct and in the form of a check list.

```
Please check the following before making a pull request:
- Make each commit title "package*: what it fixes or adds"
-- Make commit tittle 50 char max, and line wrap commit body at 72.
-- Confirm your github account is attached to each commit (your personal github icon)
-- Use a unique branch for each pull request so like "package_feature."
- Make the pull request title "package*: summary of combined effect of commits"
-- Describe the problem solved or enhancement in the pull body
-- Describe the compile environment(s) you used in the pull body
-- Describe the test environment(s) you used in the pull body
- Seek additional guidance for developer submissions at: 
-- https://www.lede-project.org/docs/guide-developer/the-source-code#general_source_structure.
-- [github contributing.md]
-- [other]

*package also may be architecture, complete device, or tools component

Thanks for your contribution.
Please remove this message before posting the pull request.
```

bobafetthotmail commented:
updated to integrate some of the feedback.

blogic commented on .github/pull_request_template:
```diff
@@ -0,0 +1,23 @@
+address
+
+- all commits must be linked to your github account (you see your logo/avatar in front of them), if they aren't 
+linked you probably need to change the local git name or email in your PC to match the one on GitHub.
```
incorrect, SoB needs to match author field.

blogic commented on .github/pull_request_template:
```diff
@@ -0,0 +1,23 @@
+- commit descriptions must describe the problem solved or enhancement in the commit body
+
+- all commits must end with "Signed-off-by: My Name <my@email.address>" where you write your real name and real email 
+address
```
s/must end with/must contain/

blogic commented on .github/pull_request_template:
```diff
@@ -0,0 +1,23 @@
+website)
+
+- commit titles start with "<package name>:" for packages "<device architecture>: " for devices
+- commit titles must state in short form  what was fixed or added
```
s/titles/title/

blogic commented on .github/pull_request_template:
```diff
@@ -0,0 +1,23 @@
+Thanks for your contribution to the LEDE project! 
+To help people review your contribution we ask you to follow some rules:
+
```
s/some/these/

blogic commented on .github/pull_request_template:
```diff
@@ -0,0 +1,23 @@
+
+
+- full patch submission rules are here 
+https://www.lede-project.org/docs/guide-developer/the-source-code#general_source_structure
```
s/rules/guidelines/

bobafetthotmail commented:
updated.

the email in the Author field for PRs opened through GitHub is the GitHub's "primary email" in the settings, I found out in this PR when you told me to fix mine https://github.com/lede-project/source/pull/471

blogic commented on .github/pull_request_template:
```diff
@@ -0,0 +1,25 @@
+address
+
+- the name used in the "Signed-off-by:" line must be the same you write in "Name" field in GitHub settings, and/or 
+see in your PC with "git config user.name"
```
still incorrect. the author field and SoB need to match.

blogic commented:
the #471 is true if and only if changes are made via the UI. if you push a locally generated tree then the author field is not changed by github.

bobafetthotmail commented:
updated. 

I kept the instructions for fixing Author field as for most newbies it is auto-generated by git (either local or on GitHub) so they have no idea of what you mean by "Author field must match SoB" unless someone explains how to fix it.

blogic commented:
coming to think of this, the patch you propose does not add a template. it adds an instruction text. a template is, as the name suggests a template that needs to be filled out, like a blank form where you just  fill in the fields.

blogic commented on .github/pull_request_template:
```diff
@@ -0,0 +1,28 @@
+
+-- if you are editing files and committing through GitHub, you must write your real name in the "Name" field in 
+   GitHub settings and the email used in the "Signed-off-by:" must be your primary github email
+
```
one of the main problems is that author and SoB do not match. adding an explanation that does not address one of the most common error does not make sense. i am not sure how many more time i need state that it is important to list this here. i have said it 4 times now. but i am happy to repeat myself even more often.

blogic commented:
adding a commit that explains how to add commits and then having that commit not follow that explanation seems odd. why is there no prefix on the subject line ?

bobafetthotmail commented:
>coming to think of this, the patch you propose does not add a template. it adds an instruction text. a template is, as the name suggests a template that needs to be filled out, like a blank form where you just fill in the fields.

The same was done by Thess for the issue template, see here https://github.com/lede-project/source/blob/master/.github/issue_template
also for OpenWRT package repo it is both a message and a template (link in first message)

Afaik it's common on github to "misuse" the template feature this way as there is no other way to post a message to people making PRs.

>adding a commit that explains how to add commits and then having that commit not follow that explanation seems odd. why is there no prefix on the subject line ?

because this is a file for github, not for a package or a device in LEDE, and the commit for the issue template did not have a prefix so I thought it was OK to not have one here too https://github.com/lede-project/source/commit/4f40f22fb31772097df95f93970aabbd63ae289e

>one of the main problems is that author and SoB do not match. adding an explanation that does not address one of the most common error does not make sense.

What? Author field is auto-generated by git by combining name and email:
On Github's git it uses what you wrote in Name (or your github nickname if no Name is set, like the case with that last guy) and Github's "primary email".
On the committer's local repo it uses user.name and user.email git settings, you can override that with --author=" myname <my.email@address>", but most guides I've seen tell to set the user.name and user.email globally.

Newbie guides never mention "author field" and there is no such thing in git settings afaik so I have to tell how to set that properly in a document supposed to help newbies.

blogic commented:
NAK from me until the note is added stating that Author field and SoB line need to match as this is one of the most common errors next to
* no prefix
* missing description
* missing SoB.

bobafetthotmail commented:
Um, I already wrote

- the Author field the must match the "Signed-off-by:" line: 

Did you miss it or is it not clear?

miltador commented:
> coming to think of this, the patch you propose does not add a template. it adds an instruction text. a template is, as the name suggests a template that needs to be filled out, like a blank form where you just fill in the fields.

Agree. This actually can be done in a style of template that user should fill instead of removing the text. As a good example of such, look at this template: https://github.com/serverless/serverless/blob/master/.github/PULL_REQUEST_TEMPLATE.md
It is structured, has points to fill and also the list of checks that can include something like "Added Signed-off-by line" and so on.

EricLuehrsen commented:
Prefix tools infrastructure with the TOOL.
`github: add pull request template/message`

blogic commented on .github/pull_request_template:
```diff
@@ -0,0 +1,28 @@
+- all commits must contain "Signed-off-by: My Name <my@email.address>" where you write your real name and real email 
+  address
+
+- the Author field the must match the "Signed-off-by:" line: 
```
missed this and looking at it now i see a trailing : that should be removed.

bobafetthotmail commented:
removed the : now.

blogic commented on .github/pull_request_template:
```diff
@@ -0,0 +1,28 @@
+
+
+Please remove this message before posting the pull request.
+
```
there is a trailing blank line

bobafetthotmail commented:
removed blank line. 

I don't see why that was an issue as this is a plain text file anyway.

blogic commented:
@bobafetthotmail because we dont want blank lines at the end of files, its a very common thing. git will actually warn you about this

blogic commented:
please also add a prefix to the subject

bobafetthotmail commented:
done, changed prefix in commit and PR.

blogic commented:
and now for the sake of completeness, we need to address the 50/72 rule. as we are addressing lots of topics we should include this aswell. could you please

* add that info the template
* make the commit follow that rule

blogic commented:
various people have pointed out that the grammar used is not optimal. furthermore why don't we just reuse existing explanations ?

http://git.kernel.org/cgit/linux/kernel/git/torvalds/linux.git/tree/Documentation/SubmittingPatches?id=HEAD#n669

bobafetthotmail commented:
> various people have pointed out that the grammar used is not optimal. 

Ok, will try harder to find grammar issues. Would be cool to state the issues you/they see. 

> furthermore why don't we just reuse existing explanations ?

Because they are huge? That's ok fit for a CONTRIBUTING.md file (again another github thing, it gets usually ignored though, see above for a link to the OpenWRT package feed's) or for a wiki page with the full submission guidelines, not for a relatively short pre-PR message.

blogic commented:
how about just using the relevant bits ? i specifically linked the paragraph that you phrased yourself

bobafetthotmail commented:
I'll see what I can do.

bobafetthotmail commented:
integrated relevant parts now.

hnyman commented on .github/pull_request_template:
```diff
@@ -0,0 +1,36 @@
+  this commit (log messages, error messages, etc.) is useful for people who might be searching the commit logs 
+  looking for a fix for their issue.  
+  If a patch fixes a compile failure, include only the part of the error log that will allow someone 
+  to find this fix. As for the commit subject, it is important to be both succinct as well as descriptive.
```
` As for the commit subject, it is important to be both succinct as well as descriptive.`
This sentence is a loose duplicate of the similar text on lines 10-11. Contextually this would fit better to lines 9-11 where commit subject is discussed, instead of here after you have been talking about the longer contents and error log extracts etc. to be included in the commit message. So, I propose to delete this sentence from here.

EricLuehrsen commented:
I dont want the message to be too wordy, but wrt. `all commits must contain "Signed-off-by: My Name <my@email.address>" where you write your real name and real email address`, we need to make sure in another doc. `contributing.md` to emphasize how critical this is to intellectual property license protection.

bobafetthotmail commented:
>we need to make sure ... to emphasize how important SoB is to intellectual property license protection.

It is currently explained in the LEDE patch submission guidelines by linking to kernel patch submission guidelines, from this link that is stated in the message https://lede-project.org/docs/guide-developer/the-source-code#submitting_patches

Is that not enough? I can place the direct link in the current message.

(btw, will also have to add to that page all the points I'm placing here, will do when this gets merged)

EricLuehrsen commented:
Sure, references are good enough as is. I got lost trying to follow and I wanted to be sure we didnt miss something. A pull message needs to be brief or no one bothers reading it.

bobafetthotmail commented:
ok removed the sentence proposed by hnyman and reworked a bit to shorten the commit description paragraph.

EricLuehrsen commented:
Did you mean commit subject should follow 50/70 rule? The body after the magical first line can be huge. Its the magical first line that is key. I am thinking of reviewing cherry-pick with `git log --oneline --grep package` for example.

bobafetthotmail commented:
added a point about "commit descriptions need to be less than 75 chars long"
more reworking and re-arranging of text to shorten it and increase readability.

stintel commented:
While we're at it, what about:
- write commit message in the imperative form: e.g. "add support for X" instead of "added support for X", would be nice if all commits were consistent, and the imperative form is shorter so it helps with the char limit
- don't capitalize first word after the prefix; it's not a full sentence so it makes no sense
- don't end the subject with a dot; same reason as above

pepe2k commented:
@stintel +3 :)

bobafetthotmail commented:
>write commit message in the imperative form

Ok.

>"don't capitalize first word after the prefix" & "don't end the subject with a dot"

These are quite frankly minor things and will go in the wiki page about guidelines, but not here as we need to list the most important things only for the sake of keeping the list short.

>avoid redundant words

Should be said already by the "be as short as possible" sentences,

>(e.g. this PR has github twice in the subject)

Ok.

mkresin commented:
> These are quite frankly minor things and will go in the wiki page about guidelines, but not here as we need to list the most important things only for the sake of keeping the list short.

As one who regularly reviews the PRs, I can say for sure it has to go into the template. From my experience only the minority of the contributor will follow links. Everything that is not mentioned in the template will be done the wrong way. And it is annoying and time consuming to fix this minor stuff before committing the patches.

bobafetthotmail commented:
sigh, ok then.

bobafetthotmail commented:
@blogic you said in another of my PRs that dma-buf (a kmod) should have the "kernel:" prefix instead of "package name:". Is that true for all kmod packages too?

blogic commented:
ideally the prefix should be linux but that is confusing so for this package we use kernel:

git log --oneline <folder>           will tell you expected prefixes

bobafetthotmail commented:
please wait, updated the PR too soon.

bobafetthotmail commented:
ok, done.

bobafetthotmail commented:
If there are no more comments on this, I think it might make sense merging it.

thagabe commented:
ping, will this be merged?

bobafetthotmail commented:
@mkresin you seem to be referencing this commit quite a bit lately, can you merge this or are we waiting for @blogic ?

stefanct commented:
The `build` prefix is missing for general build changes that are not targeted at something in `toolchain/` AFAICT.

bobafetthotmail commented:
@stefanct I don't understand what you are saying. Can you explain it better?

hnyman commented:
> >The `build` prefix is missing for general build changes that are not targeted at something in toolchain/ AFAICT.

>  I don't understand what you are saying. Can you explain it better? What you want me to do?

He suggests that the prefix examples list should include also "build:" for commits like https://github.com/lede-project/source/commit/f938de791438fc932c5a5ccfecd8f33050e3e103 and https://github.com/lede-project/source/commit/fe1e3622a271386ea0413c97d9884e9935e17f11 that target the build logic itself. Also "toolchain:" has been used in some similar cases, e.g. in https://github.com/lede-project/source/commit/b2c66724d71aa252b2d21c9984e7866a025ffd86

stefanct commented:
@bobafetthotmail You are listing some possible prefixes here: https://github.com/lede-project/source/pull/592/commits/c7cfdff4fa40f53511b8c8a4bd2e60aa980dc6fd#diff-f7162af36e86259e13bfbf7f0fc3eb48R11
I was suggesting that you add some more... what @hnyman said while I was typing ;)
However, `toolchain:` and `build:` seem to be distinct to me (there are build files outside of `toolchain/`.

bobafetthotmail commented:
ok updated with the "build:" option.
I think a "toolchain" option is unnecessary as there is a "tool name" option already that will catch the tools in /toolchain, but this isn't my call anyway.

pepe2k commented:
Hello @bobafetthotmail,
 
I would like to move on with this one.
I spoke with @mkresin about the idea of having GitHub PR template inside the repository and few things came out:
 
- we agreed that we need this guideline for submitters, but more like a general submission guideline, not only for GitHub PR
- any kind of change in this template text would require a separate commit which we would like to avoid (next point explains why)
- we don't keep documentation with sources anymore, so the best place for this would be Wiki (atm, docs/guide-developer/ looks like a good candidate for it, but feel free to suggest something else)
- we both think that it should be somehow highlighted and easily available in Wiki menu (side navigation), not "hidden" like the "Submitting Patches" section now
- when this will be available in Wiki, we can rework this template and instead of whole guideline text, include only short information and link to the Wiki so that in case of any changes in guideline, we won't need to make changes in code repository

Cheers,
Piotr

bobafetthotmail commented:
Hm, ok for me. Will migrate bulk of the submission requirements to the wiki and make the required changes in a few days.

bobafetthotmail commented:
Updated, the wiki article is here https://lede-project.org/docs/guide-developer/contribution (it is accessible from "Contributor's Guide" link on the sidebar on the left)

mkresin commented:
Excellent work so far.

Any opinions on renaming:

- "contributers guide" to "Submitting patches"
- "Bug reports" to "Reporting bugs"

EricLuehrsen commented:
I would like to see a simple form underneath like OpenWrt packages, but this this has been run out. Lets just merge this simple piece of literature and be done with it.

EricLuehrsen commented:
I would like to see a simple form underneath like OpenWrt packages, but this this has been run to long. Lets just merge this simple piece of literature as is now  and be done with it.

bobafetthotmail commented:
changed the titles on wiki as suggested.

pepe2k commented:
@bobafetthotmail thank you for work on this!

One thing:
- redable -> readable (typo?)

Also, it's probably not related to the PR that much, but I will just drop it here.

I have some problem with the Wiki left navigation menu: when I'm in `Submitting patches` section, the `Documentation` menu is expanded. Is it some kind of bug/mistake in the menu/pages hierarchy? This is how it looks now:

[![2017-03-06_090115.jpg](https://s13.postimg.org/jy74ux7qf/2017_03_06_090115.jpg)](https://postimg.org/image/x2cp7lzs3/)

Cheers,
Piotr

bobafetthotmail commented:
> redable -> readable (typo?)

Sigh, respinned the patch again, thanks.

> I have some problem with the Wiki left navigation menu: 

That's because the page is (was) technically in Developer Guide "namespace thing" in the wiki, see the (old) URL? 
https://lede-project.org/docs/guide-developer/contribution
So the wiki expands that menu like when you are in any other article placed under quickstart, user guide or developer guide.

Since I had to fix that typo anyway, I also moved the page to root wiki "namespace" (and changed the link accordingly) so it won't open the menu like that.

pepe2k commented:
@mkresin, @blogic, what do you think about thist now?

mkresin commented:
@pepe2k fine with me. Feel free to merge the PR and hopefully it will save us some work.

pepe2k commented:
@bobafetthotmail and all others involved in work on this, thank you! I have merged this into [my staging tree](https://git.lede-project.org/?p=lede/pepe2k/staging.git;a=summary).
